service/matching/matcher_test.go

```diff
@@ -267,24 +267,51 @@ func (t *MatcherTestSuite) TestRejectSyncMatchWhenBacklog() {
 func (t *MatcherTestSuite) TestForwardingWhenBacklogIsYoung() {
 	historyTask := newInternalTaskForSyncMatch(randomTaskInfo().Data, nil)
 
-	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	ctx, cancel := context.WithTimeout(context.Background(), 4*time.Second)
 	intruptC := make(chan struct{})
 
-	// poll forwarding attempt happens when there is no backlog
-	t.client.EXPECT().PollWorkflowTaskQueue(gomock.Any(), gomock.Any(), gomock.Any()).Return(&matchingservice.PollWorkflowTaskQueueResponse{}, errMatchingHostThrottleTest)
-	go t.childMatcher.Poll(ctx, &pollMetadata{}) //nolint:errcheck
-	time.Sleep(time.Millisecond)
+	var wg sync.WaitGroup
+	wg.Add(1)
 
-	// task is not forwarded because there is a poller waiting
-	youngBacklogTask := newInternalTaskFromBacklog(randomTaskInfoWithAge(time.Second), nil)
+	t.client.EXPECT().PollWorkflowTaskQueue(gomock.Any(), gomock.Any(), gomock.Any()).Do(
+		func(arg0 context.Context, arg1 *matchingservice.PollWorkflowTaskQueueRequest, arg2 ...interface{}) {
+			wg.Done()
+		},
+	).Return(&matchingservice.PollWorkflowTaskQueueResponse{}, errMatchingHostThrottleTest)
+
+	go func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 4*time.Second)
+		// poll forwarding attempt happens when there is no backlog
+		_, err := t.childMatcher.Poll(ctx, &pollMetadata{})
+		t.Assert().NoError(err)
+		cancel()
+	}()
+	// This ensures that the poll request has been forwarded to the parent partition before the offer is made.
+	// Without this, there is a chance that the request is matched on the child partition, which will fail the test by
+	// complaining about a missing PollWorkflowTaskQueue.
+	wg.Wait()
+
+	// to ensure poller is now blocked locally
+	time.Sleep(2 * time.Millisecond)
```

Comment:
similarly, does 1 millisecond work?

I might be nitpicking here, but my perspective is that if possible we should avoid "magic" constants where if in a year someone reads this, they won't be able to know why this is 2ms instead of 1ms, or why the context timeouts were 4 seconds instead of 1 second.

Reply:
I totally agree - my thoughts on this:

The jump from 1 second to 4 second, for each test run, is a lot and I have changed it since it works on a context timeout of 1 second.

The jump from 1 millisecond to 2 milliseconds is not huge for a single test run. I tested it out on 1 millisecond (ran it about 15000 times) and it works. However, I think it's a good idea to be a little conservative here and go with 2 milliseconds. Here's the chain of execution which makes me reason with this:
- The poller attempts to forward and we trigger a `wg.Done()`. After returning from the forwarding function, where it has received an error, the poller begins to poll locally.
- This unblocks the main goroutine and we trigger `time.Sleep(1 * time.Millisecond)`
- The main go-routine calls `MustOffer` which expects the poller to be polling locally.

My fear, maybe irrational, is that there could be a chance that 1 Millisecond might be too less for the poller to *begin* polling locally when we offer the task. I think two milliseconds is being conservative to the point where there is not *much* of an additional delay.

The total test delay now seems to be - 1 x 2 milli + 1 x 1 milli + delay from waitGroups (< 1 second)

wdyt?

Reply:
I wrote a long comment trying to think through it, but ended up confusing myself, then just tested it. 1ms sleep flakes if run 100,000x whereas 2ms sleep does not. I think 2ms is good :)

Reply:
Just something to think about: running hundreds of thousands of times locally is great, but a lot of time-based flakiness won't show up on your (relatively quite powerful) laptop at all, but will show up in CI on relatively weak VMs and much more cpu contention. So if you're using local tests to validate flakiness, maybe try running a cpu stress test in the background as well.